tests/integration/v3_grpc_test.go

```diff
@@ -1952,3 +1953,77 @@ func waitForRestart(t *testing.T, kvc pb.KVClient) {
 		t.Fatalf("timed out waiting for restart: %v", err)
 	}
 }
+
+func TestV3ReadonlyTxnCancelledContext(t *testing.T) {
+	integration.BeforeTest(t)
+	clus := integration.NewCluster(t, &integration.ClusterConfig{
+		Size: 1,
+		// Context should be cancelled on the second check that happens inside rangeKeys
+		GrpcInterceptors: []grpc.UnaryServerInterceptor{injectMockContextForTxn(newMockContext(2))},
```

Comment:
The `2` here is a little tricky. 

I suggest to simplify the test case for better readability:
1. Remove the Put (line 1968), and let's test the `readonly transaction` only;
2. always return a closed channel in `(ctx *mockCtx) Done();`

Reply:
Agree that counting calls on context to trigger reaction is not maintainable. 
Thinking about alternatives, how about intercepting zap logs, using zap.hooks: 

https://stackoverflow.com/questions/64803634/how-to-add-a-hook-into-a-zap-logger

Still not perfect, but allows to e.g. disable context when specific transaction is debug-logged.

Reply:
@ahrtr Put is there to prevent short circuit in storeTxnRead.Range. We need some data to trigger `case <-ctx.Done():`
Always returning closed channel doesn't work because it fails early in `EtcdServer.linearizableReadNotify` and never gets to `txn.Txn`

@ptabor how do you trigger `context cancelled` inside `storeTxnRead.Range` if you don't let the first check in `EtcdServer.linearizableReadNotify` to pass? Maybe I'm missing something, but how will intercepting zap logs help? As far as I can tell only failure inside storeTxnRead.Range triggers panic, so that's what we want to model in the test.

Another approach I see is to add test only on txn.Txn level and not replicate panic as integration test. Maybe not ideal, but integrations tests don't have to cover every failure scenario.